File: Day<08>/Solution.cpp
#include<bits/stdc++.h> 
using namespace std; 

 // } Driver Code Ends

class Solution{
    
    public:
    int maxCandy(int height[], int n) 
    { 
    
    	// To store the maximum water so far 
    	int maximum = 0; 
    
    	// Both the pointers are pointing at the first 
    	// and the last buildings respectively 
    	int i = 0, j = n - 1; 
    
    	// While the water can be stored between 
    	// the currently chosen buildings 
    	while (i < j) 
    	{ 
    
    		// Update maximum water so far and increment i 
    		if (height[i] < height[j]) 
    		{ 
    			maximum = max(maximum, (j - i - 1) * height[i]); 
    			i++; 
    		} 
    
    		// Update maximum water so far and decrement j 
    		else if (height[j] < height[i]) 
    		{ 
    			maximum = max(maximum, (j - i - 1) * height[j]); 
    			j--; 
    		} 
    
    		// Any of the pointers can be updated (or both) 
    		else
    		{ 
    			maximum = max(maximum, (j - i - 1) * height[i]); 
    			i++; 
    			j--; 
    		} 
    	} 
    
    	return maximum; 
    }   
};

// { Driver Code Starts.

int main(){
    int t;
    cin>>t;
    while(t--)
    {
    	int n;
    	cin>>n;
    	int height[n];
    	for (int i = 0; i < n; ++i)
    	{
    		cin>>height[i];
    	}
    	
    	Solution obj;
    	
    	cout << obj.maxCandy(height,n)<<"\n";
    }
    return 0;
}

File: Day<08>/Solution.py
class Solution:
    def maxCandy(self, height, n):
    	# To store the maximum water so far
    	maximum = 0;

    	# Both the pointers are pointing at the first and the last buildings respectively
    	i = 0
    	j = n - 1

    	# While the water can be stored between the currently chosen buildings 
    	while (i < j):
    		# Update maximum water so far and increment i
    		if (height[i] < height[j]):
    			maximum = max(maximum, (j - i - 1) * height[i]);
    			i += 1;

    		# Update maximum water so far and decrement j
    		elif (height[j] < height[i]):
    			maximum = max(maximum, (j - i - 1) * height[j]);
    			j -= 1;

    		# Any of the pointers can be updated (or both)
    		else:
    			maximum = max(maximum, (j - i - 1) * height[i]);
    			i += 1;
    			j -= 1;

    	return maximum;

if __name__ == '__main__':
    t = int(input())
    for _ in range(t):
        n = int(input())
        arr = list(map(int, input().strip().split()))
        obj = Solution()
        print(obj.maxCandy(arr,n))
